fix(schema): Fix default value change detection for Time fields

cogk opened the pull request:
Make the handling of default values for decimal column coherent with the handling for non-decimal columns.
This helps avoid `ValueError: could not convert string to float: 'NULL'`
_I've encountered the error but is it really a real error that happens outside of a test env?_

https://github.com/frappe/frappe/blob/a6e87dad27fa4c9b278db41b7bcfe1955805dcdc/frappe/database/schema.py#L319-L320

# Stack trace

<details>

<summary>Stack trace triggered by creating a custom field (SLA fields) while running tests in a custom app</summary>

```py
  File "/apps/frappe/frappe/model/document.py", line 1185, in fn
    return method_object(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    args = ()
    kwargs = {}
    method = 'on_update'
    method_object = <bound method CustomField.on_update of <CustomField: doctype=Custom Field Lead-service_level_section>>
    self = <CustomField: doctype=Custom Field Lead-service_level_section>
  File "/apps/frappe/frappe/custom/doctype/custom_field/custom_field.py", line 214, in on_update
    frappe.db.updatedb(self.dt)
    self = <CustomField: doctype=Custom Field Lead-service_level_section>
    validate_fields_for_doctype = <function validate_fields_for_doctype>
  File "/apps/frappe/frappe/database/mariadb/mysqlclient.py", line 492, in updatedb
    db_table.sync()
    db_table = <frappe.database.mariadb.schema.MariaDBTable object>
    doctype = 'Lead'
    meta = None
    res = ((0,),)
    self = <frappe.database.mariadb.mysqlclient.MariaDBDatabase object>
  File "/apps/frappe/frappe/database/schema.py", line 48, in sync
    self.alter()
    self = <frappe.database.mariadb.schema.MariaDBTable object>
  File "/apps/frappe/frappe/database/mariadb/schema.py", line 74, in alter
    col.build_for_alter_table(self.current_columns.get(col.fieldname.lower()))
    col = <frappe.database.schema.DbColumn object>
    self = <frappe.database.mariadb.schema.MariaDBTable object>
  File "/apps/frappe/frappe/database/schema.py", line 308, in build_for_alter_table
    self.default_changed(current_def)
    column_type = 'decimal(21,9)'
    current_def = {'name': 'total_hold_time', 'type': 'decimal(21,9)', 'default': 'NULL', 'index': 0, 'unique': 0, 'not_nullable': 0}
    self = <frappe.database.schema.DbColumn object>
  File "/apps/frappe/frappe/database/schema.py", line 331, in default_changed
    return self.default_changed_for_decimal(current_def)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    current_def = {'name': 'total_hold_time', 'type': 'decimal(21,9)', 'default': 'NULL', 'index': 0, 'unique': 0, 'not_nullable': 0}
    self = <frappe.database.schema.DbColumn object>
  File "/apps/frappe/frappe/database/schema.py", line 358, in default_changed_for_decimal
    and float(current_def["default"]) == 0.0
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    current_def = {'name': 'total_hold_time', 'type': 'decimal(21,9)', 'default': 'NULL', 'index': 0, 'unique': 0, 'not_nullable': 0}
    self = <frappe.database.schema.DbColumn object>
ValueError: could not convert string to float: 'NULL'
```

</details>

cogk committed:
fix(schema): Handle "NULL" in default_changed_for_decimal

cogk commented:
@ankush Hey :wave:, how does the comparison work for `time` fields?

There is no `Time` field in Frappe so this a _test blind spot_ I believe.

From my own tests I have the following failures (extra migrations) because of "default changed":

| name               | type    | cur_default       | new_default | 
|--------|--------|--------|--------|
| daily_maximum_time | time(6) | '21:00:00.000000' | 21:00 |
| daily_minimum_time | time(6) | '06:00:00.000000' | 06:00 |
| from_time          | time(6) | '09:00:00.000000' | 9:00:00 |
| pickup_from        | time(6) | '09:00:00.000000' | 09:00 |
| pickup_to          | time(6) | '17:00:00.000000' | 17:00 |
| send_at_time       | time(6) | '09:00:00.000000' | 09:00:00 |
| to_time            | time(6) | '17:00:00.000000' | 17:00:00 |

ankush commented:
> how does the comparison work for time fields?

It's broken :smile: But only 2 rare places in codebase seem to use default time, so left it :woozy_face:

ankush commented:
Casting is the write way to fix it, just need to avoid fallbacks that are present in all casting functions :facepalm: Can use python stdlib for parsing and throw it away if it's not parseable.

ankush commented:
Your PR description is probably outdated? I fixed that code on same day.

cogk commented:
Yeah my PR description is bad :laughing:, however the code works I just tested it :nerd_face:

---

By the way congrats on the performance work :clap:, I have test pipelines that are two times faster now which is incredible :shaking_face: (this is life-changing)

cogk committed:
fix(schema): Fix default value change detection for Time fields

cogk committed:
chore: format

ankush committed:
fix: Don't set zero default during change

This makes the check more aggresive. We by default set 0 on all numeric
fields, so we should ideally only migrate default between numbers and
not check `NULL` ever.
